### ISSUE TITLE
Allow "-" in filenames

### ISSUE THREAD

**Issue by robyngit** [NCEAS/metacatui#2228]
Files uploaded via the editor are sometimes renamed. All characters in filenames that are not letters or digits are replaced with an underscore. This is too conservative - we should allow "-" in filenames as well.

**Comment by mbjones** [comment]
We should try to not rename if possible, and only rename if the characters are truly illegal. Its a problem, for example, if the filename is referenced in an R or python script, or elsewhere in the metadata text, and it doesn't match up. Maybe we are being too greedy in the characters we eliminate?

**Comment by robyngit** [comment]
@mbjones agree that should aim to minimize renaming and only replace characters that are actually illegal or problematic across different platforms. I propose we come up with a set of allowable characters. Here's what I've got so far, input appreciated!!

**Characters to Allow:** Universally legal
1. Alphanumeric Characters `A-Z, a-z, 0-9`
2. Hyphen `-`
3. Underscore `_`
4. Period `.`

**Characters for Discussion:** These characters can cause issues in some contexts
1. Space ` `
1. Parentheses `()`, Square Brackets `[]`
2. Exclamation Mark `!`, At Symbol `@`, Number Sign `#`, Dollar Sign `$`, Percent Sign `%`
3. Plus Sign `+`, Equal Sign `=`
4. Comma `,`
5. Single Quotes `'`
6. Other characters...?

**Comment by mbjones** [comment]
What about all other non-ascii characters, in the Unicode range?

**Comment by mburrus** [comment]
Hi @robyngit I just wanted to point out that ESS-DIVE recently received a use case related to this issue ticket. Recording here for tracking purposes.

This user requested several times that we allow hyphens and periods to be included in their file name. The file she wants to upload is named `00001-02715.00001-01803` however it gets renamed to `00001_02715.00001-01803`. See quoted response below, listed in chronological order that we received them: 

> It is important that the dataset actually be named: 00001-02715.00001-01803 with no underscores because the filename must be in that format for WRF to read.

>  The 00001_02715.00001-01803 file (although the first underscore in this filename should have been a dash, I’m not sure why that transferred differently).

> Thanks also for looking into the naming of the binary file. I tried several times to rename it manually, but I was unsuccessful. The filename must have dashes and not underscores on both sides of the dot or it will not work. 

As an FYI, I conferred with one of our team members who regularly conducts metadata and file reviews, and he said he most often sees `-` and `.` characters used in file names. He didn't recall seeing other characters used. Additionally, he said this bug effects reporting format datasets using File Level Metadata (FLMD) files 
because data contributors would fill out the FLMD files with a file name that was different from the file name in our system.  

Adding @vchendrix and a reference to the Jira ticket for tracking purposes: [EDSUPPORT-2670](https://ess-dive.atlassian.net/browse/EDSUPPORT-2670).

**Comment by mbjones** [comment]
I agree. We should not be renaming files in any way. Any idea of why we do this? It's not needed on the backend.

**Comment by dschristianson** [comment]
Hi @robyngit. Adding support for this issue to be tackled. I work on a project with @vchendrix that uses MetacatUI. The renaming of files is affecting our users as well, especially hyphen and period. Accepting the list of allowable characters above (alphanumeric, underscore, hyphen, period) would be great. Thanks!

**Comment by mbjones** [comment]
Thanks @dschristianson for the ping.

@robyngit Picking up our conversation from last May, any idea of why we do this? It's not needed on the backend because files are stored using their checksum, not their filename. The filename is in the system metadata doc, and then used to suggest a filename on download. I don't see any need for character restrictions, or an approved characters list. The main issue you'll need to check is that filenames are properly escape when saved in their XML home in SystemMetadata, but that should be automatic in the sysmeta serializer/deserializer.

**Comment by robyngit** [comment]
@mbjones after sifting through commit & issue history, I'm not completely sure WHY this decision was made. Here's what I did find out:

The change to file naming convention was made in [May 2019](https://github.com/NCEAS/metacatui/commit/26943c0596a34612cb5545cf07b696ba7594c305), after a some [discussion](https://github.com/NCEAS/metacatui/issues/915) that happened over a few months:

> We discussed replacing all non a-zA-z0-9 characters with underscores, which includes hyphens. 

> hyphens, along with all other non-alphanumeric characters in the UUID, replaced with underscores

> I don't recall discussing the naming convention you mentioned for EMLs

> It sounds like we all agree that we should replace the spaces with underscores

> We decided to change the metadata file names to use the title only, replacing all nonalphanumeric characters with underscores, and truncating after the first word break before 50 characters. Multiple underscores in a row would be consolidated down to one underscore.

It appears the filename issue was first brought up in during an arctica meeting in [Feb 2019](https://docs.google.com/document/d/1WbB_GRCo-hivgBintHxr-aMwEpecMZhu-YSDngiLRtM/edit?tab=t.0#heading=h.6qglz6img07c) in reference to "NCEI replication". Discussion continued [later in the month](https://docs.google.com/document/d/1WbB_GRCo-hivgBintHxr-aMwEpecMZhu-YSDngiLRtM/edit?tab=t.0#heading=h.qg4l5go9u2tu):

> (Jesse) In fileNames, {UUID-for-RDF} replace all non [A-Za-z0-9]with _ (underscore)

It seems like this was a long, thoroughly discussed choice, though the exact rationale for it isn't clear. I wouldn't want to undo something that may have been a fix to some other problem. Does any of the above ring a bell @mbjones ?

**Comment by mbjones** [comment]
Yes, it rings a bell now due to the NCEI requirements from NOAA. I read through that whole conversation, and I did not see the logic for the changes, other than following NCEI rules. Let's discuss. We really shouldn't be renaming files, and I am sure it is surprising to some researchers. We do need to choose names for the two files we create (for the RM and EML doc), and I think the approach we took for those is fine. But data files should be named by the user. At most, we might want to recommend a change in those names if they don't have a consistent or logical convention. And while it is best practice to not use spaces, and certainly not non-printing characters, we certainly will encounter them in some places, and there are probably legitimate reasons to use them. Our renaming pattern is draconian.

**Comment by mbjones** [comment]
Thanks @robyngit great summary and glad to see progress here. 

I think we should have a quick chat about this with a few of our curation team and a few others (@justinkadi @maggieklope @jeanetteclark @vchendrix ) before we decide the rules to implement. The Windows restrictions in particular are onerous on other platforms, and I wouldn't want to prevent use of some of those characters. In particular, I think we should make as few (or even no) automatic changes to people's filenames, as they very much may depend on those exact paths in scripts or other references. If a filename is illegal, maybe we should flag it in the UI and force the user to change it, rather than changing it ourselves?

I think a another way to deal with these issues is on download -- if a browser or other user agent from a Windows OS requests a file, can't we then modify the filename to be compliant with that filesystem? Do we even need to? As there are multiple sources of filenames (e.g., uploading packages from R, python, curl, etc), MetacatUI must be prepared to deal with filenames that are legal on one system but not on another when handling download requests. Maybe the browser already handles this anyways, in that all we are doing is setting the name in the `Content-Disposition` header, and the browser actually writes it to disk, right? What do browsers do with filenames that have illegal headers in their content metadata?

That said, I think the standard set of exclusions that are illegal on all platforms (control characters and slash, for example) can be stripped, but we should let the user know we changed the name and why.
